xl-v1-healing.go

```diff
@@ -16,33 +16,90 @@
 
 package main
 
-import "sync"
-
-// Get the highest integer from a given integer slice.
-func highestInt(intSlice []int64, highestInt int64) (highestInteger int64) {
-	highestInteger = highestInt
-	for _, integer := range intSlice {
-		if highestInteger < integer {
-			highestInteger = integer
-			break
+import (
+	"sort"
+	"sync"
+)
+
+// commonUUID returns a maximally occurring UUID from a list of UUIDs.
+func commonUUID(uuids []string) (uuid string) {
+	// Variables.
+	var (
+		pIndex     int    // Holds the value of th first non-sentinel element.
+		prevUUID   string // Holds the value of the first non-sentinel element.
+		commonUUID string // Holds the final value of uuid most occurred.
+		count      int    // Counter for same occurrence of elements.
+		maxima     int    // Counter for remembering max occurrence of elements.
+	)
+	// Sort so that the sentinel value are staggerred at the beginning.
+	sort.Strings(uuids)
+	for pIndex, prevUUID = range uuids {
+		// UUID is a sentinel value ignore it.
+		if prevUUID == uuidSentinel {
+			continue
+		}
+		break
+	}
+
+	// Ignore uuids upto the previous index,
+	// based on the last known sentinel value.
+	uuids = uuids[pIndex:]
+
+	// Find the common cardinality, by counting.
+	for _, u := range uuids {
+		if u != prevUUID {
+			// Current maxima is lesser than occurrence of elements.
+			// Reset maxima to new maxima and the previous element
+			// is the maximum element.
+			if maxima < count {
+				maxima = count
+				commonUUID = prevUUID
+			}
+			count = 0
 		}
+		count++
+		prevUUID = u
 	}
-	return highestInteger
+
+	// Eventually if we see the occurrence of a certain element
+	// is higher than maxima, we return the last element.
+	if count > maxima {
+		commonUUID = prevUUID
+	} // else for all other cases we return the choosen common element.
+	return commonUUID
+}
```

Comment:
It was hard to read the logic, easier would be to use map (though map will be inefficient than slice):
```
func commonUUID(uuids []string) (uuid string) {                                                                                                                                 
        uuidMap := make(map[string]int)                                                                                                                                         
        for _, uuid := range uuids {                                                                                                                                            
                if uuid == uuidSentinel {                                                                                                                                       
                        continue                                                                                                                                                
                }                                                                                                                                                               
                uuidMap[uuid]++                                                                                                                                                 
        }                                                                                                                                                                       
        commonUUID := ""                                                                                                                                                        
        maxCount := 0                                                                                                                                                           
        for uuid, count := range uuidMap {                                                                                                                                      
                if count > maxCount {                                                                                                                                           
                        maxCount = count                                                                                                                                        
                        commonUUID = uuid                                                                                                                                       
                }                                                                                                                                                               
        }                                                                                                                                                                       
        return commonUUID                                                                                                                                                       
}                           
```